src/swaps/lifi/LifiSwapProvider.ts

```diff
@@ -0,0 +1,321 @@
+      console.warn(e);
+      return null;
+    }
+  }
+
+  async newSwap(swap: SwapRequest<LifiSwapHistoryItem>) {
+    const approvalAddress = swap.quote.bestRoute?.estimate.approvalAddress;
+    const approveTx = await this.approve(swap, false, approvalAddress);
+    const updates = approveTx || (await this.initiateSwap(swap));
+
+    return { id: uuidv4(), ...updates };
+  }
+
+  private async initiateSwap({ network, walletId, quote }: SwapRequest<LifiSwapHistoryItem>) {
+    const client = super.getClient(network, walletId, quote.from, quote.fromAccountId);
+    const txData: TransactionRequest = quote.bestRoute?.transactionRequest as TransactionRequest;
+    const fromFundTx = await client.wallet.sendTransaction(txData);
+
+    return {
+      status: 'WAITING_FOR_INITIATION_CONFIRMATIONS',
+      fromFundTx,
+      fromFundHash: fromFundTx.hash,
+    };
+  }
+
+  //  ======== FEES ========
+  async estimateFees(feeRequest: EstimateFeeRequest<string, LifiSwapHistoryItem>) {
+    if (feeRequest.txType in this._txTypes()) {
+      const fees: EstimateFeeResponse = {};
+      const gasCost = (feeRequest.quote.bestRoute?.estimate.gasCosts as GasCost[])[0];
+
+      const approvalData = await this.buildApprovalTx(
+        feeRequest,
+        false,
+        feeRequest.quote.bestRoute?.estimate.approvalAddress
+      );
+
+      let approvalGas = 0;
+      if (approvalData) {
+        const client = this.getClient(
+          feeRequest.network,
+          feeRequest.walletId,
+          feeRequest.quote.from,
+          feeRequest.quote.fromAccountId
+        );
+        approvalGas = (await client.chain.getProvider().estimateGas(approvalData)).toNumber();
+      }
+
+      const nativeAsset = chains[cryptoassets[feeRequest.quote.from].chain].nativeAsset;
+
+      // have only one fee and it cannot be changed by user
+      for (const feePrice of feeRequest.feePrices) {
+        const gasPrice = new BN(feePrice).times(1e9); // ETH fee price is in gwei
+        const approvalFee = new BN(approvalGas).times(gasPrice);
+        fees[feePrice] = unitToCurrency(cryptoassets[nativeAsset], approvalFee.plus(gasCost.amount));
+      }
+
+      return fees;
+    }
+
+    return null;
+  }
+
+  async getMin(_quoteRequest: QuoteRequest) {
+    return new BN(0);
+  }
+
+  // ======== HELPERS ========
+
+  private getDefaultRouteOptions() {
+    return {
+      slippage: 5 / 100, // 5%
+      order: Orders[0], // 'RECOMMENDED'
+    };
+  }
+
+  // cross chain swaps info can be aquire only via Li.Fi API
+  private async getCrossChainSwapStatus(quote: LifiSwapHistoryItem) {
+    const result = await this._httpClient.nodeGet('/status', {
+      bridge: quote.bestRoute?.tool,
+      fromChain: this.getChainByID(quote.bestRoute?.action.fromChainId as number),
+      toChain: this.getChainByID(quote.bestRoute?.action.toChainId as number),
+      txHash: quote.fromFundHash,
+    });
+
+    return result;
+  }
+
+  private getChainByID(id: number) {
```

Comment:
Can you explain what this does? Not very clear. There's a lot of casting between where it's called and inside this function and it's making me worried because the types are not being enforced at all

Reply:
This method takes ChainID and returns the key of this enum: https://github.com/lifinance/types/blob/3ba8bb7712adcd87a2023f655cd6c1f2c48b253e/src/base.ts